streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinNode.java

```diff
@@ -0,0 +1,154 @@
+    }
+
+    static final class KTableKTableJoinNodeBuilder<K, V, V1, V2, VR> {
+
+        private String name;
+        private String predecessorNodeName;
+        private String[] joinThisStoreNames;
+        private ProcessorParameters<K, V1> joinThisProcessorParameters;
+        private String[] joinOtherStoreNames;
+        private ProcessorParameters<K, V2> joinOtherProcessorParameters;
+        private ProcessorParameters<K, VR> joinMergeProcessorParameters;
+        private ValueJoiner<? super V, ? super V1, ? extends V2> valueJoiner;
+        private String thisJoinSide;
+        private String otherJoinSide;
+
+        private KTableKTableJoinNodeBuilder() {
+        }
+
+        KTableKTableJoinNodeBuilder<K, V, V1, V2, VR>  withJoinThisStoreNames(final String[] joinThisStoreNames) {
+            this.joinThisStoreNames = joinThisStoreNames;
+            return this;
+        }
+
+        KTableKTableJoinNodeBuilder<K, V, V1, V2, VR>  withJoinThisProcessorParameters(final ProcessorParameters<K, V1> joinThisProcessorParameters) {
+            this.joinThisProcessorParameters = joinThisProcessorParameters;
+            return this;
+        }
+
+        KTableKTableJoinNodeBuilder<K, V, V1, V2, VR>  withName(String name) {
+            this.name = name;
+            return this;
+        }
+
+        KTableKTableJoinNodeBuilder<K, V, V1, V2, VR>  withJoinOtherStoreNames(final String[] joinOtherStoreNames) {
+            this.joinOtherStoreNames = joinOtherStoreNames;
+            return this;
+        }
+
+        KTableKTableJoinNodeBuilder<K, V, V1, V2, VR>  withPredecessorNodeName(final String predecessorNodeName) {
+            this.predecessorNodeName = predecessorNodeName;
+            return this;
+        }
+
+        KTableKTableJoinNodeBuilder<K, V, V1, V2, VR>  withJoinOtherProcessorParameters(final ProcessorParameters<K, V2> joinOtherProcessorParameters) {
+            this.joinOtherProcessorParameters = joinOtherProcessorParameters;
+            return this;
+        }
+
+        KTableKTableJoinNodeBuilder<K, V, V1, V2, VR>  withJoinMergeProcessorParameters(final ProcessorParameters<K, VR> joinMergeProcessorParameters) {
+            this.joinMergeProcessorParameters = joinMergeProcessorParameters;
+            return this;
+        }
+
+        KTableKTableJoinNodeBuilder<K, V, V1, V2, VR>  withValueJoiner(final ValueJoiner<? super V, ? super V1, ? extends V2> valueJoiner) {
+            this.valueJoiner = valueJoiner;
+            return this;
+        }
+
+        KTableKTableJoinNodeBuilder<K, V, V1, V2, VR>  withThisJoinSide(final String thisJoinSide) {
+            this.thisJoinSide = thisJoinSide;
+            return this;
+        }
+
+        KTableKTableJoinNodeBuilder<K, V, V1, V2, VR>  withOtherJoinSide(final String otherJoinSide) {
+            this.otherJoinSide = otherJoinSide;
+            return this;
+        }
+
+        KTableKTableJoinNode<K, V, V1, V2, VR> build() {
```

Comment:
Again, not a change suggestion comment: I'm wondering if the `build()` call of this class will only be called once in its life time? If yes, we could just incorporate it into the `KTableKTableJoinNode` itself with a no-arg constructor, and allow all fields to be not-final. This is correlated to my other comment above.

We can discuss about this more in the follow-up PR.

Reply:
Fair enough.